### DISCUSSION TITLE
Shape render for bagpart

### DISCUSSION THREAD

**Discussion by sayyad84** [OrchardCMS/OrchardCore#11278] [Q&A]
Hi,
How can I access specific widget from flow widget by widget name or index? I have multiple widget in flow and I need access in different places of template not with {% for %}.
Thanks.

**Comment by ns8482e** [comment]
You can customize and design `BagPart` template with alternates the way you need . 
- Use  [console_log](https://docs.orchardcore.net/en/latest/docs/reference/modules/Contents/#logging-to-the-browser-console) to see Shape properties
- Use [shape_build_display](https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#shape_build_display) to create a shape for content item of BagPart
- Use [shape_render](https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#shape_render) to render the shape at place you desire within template you customize

**Reply by dodyg** [reply]
@ns8482e can you please elaborate for on this?

![content-item](https://user-images.githubusercontent.com/1761249/195998241-7eae4080-ca17-4df3-8ad4-59625197ea7e.png)

For example I am using a flow part and I have no idea how to accomplish picking specific template to render a specific content type inside a flow.

`{{ Model.ContentItem.Content.FlowPart.Widgets[0] }}` 

How can I tell Liquid to render this particular shape with a specific alternate?

**Reply by ns8482e** [reply]
Model.ContentItem.Content.FlowPart.Widgets[0] is not a shape it’s json of widget content item. So you need to build shape using [shape_build_display](https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#shape_build_display)

**Reply by dodyg** [reply]
I am lost

Here I am trying to assign the `FlowPart` shape with a specific alternate

![problem-2](https://user-images.githubusercontent.com/1761249/196042949-218be169-9eff-4d6d-83d6-08127f3b890f.png)

This is `{{ Model.Content | console_log}}` displaying the structure of the shape. The alternate for the `FlowPart` isn't set.

![problem-](https://user-images.githubusercontent.com/1761249/196042927-062e27bc-f5ea-4a5a-ad09-b37dd38d6faf.png)

**Reply by ns8482e** [reply]
I don’t think Model.Content.Items properly is accessible from liquid

use name of shape ie Model.Content.FlowPart

**Reply by dodyg** [reply]
Thanks `Model.Content.FlowPart` works!

![main](https://user-images.githubusercontent.com/1761249/196122691-672a952f-a6a7-40ac-87af-d7fa30be088c.png)
![main-2](https://user-images.githubusercontent.com/1761249/196122697-ddcc196f-6370-428d-92a6-b031a2b6433e.png)

So now `Model.Content.FlowPart` has 4 alternates. How do I tell it to use a specific alternate?

I thought I could use `shape_clear_alternates` then use `shape_add_alternates`, **the problem is that `shape_clear_alternates` doesn't work**.

![clear-1](https://user-images.githubusercontent.com/1761249/196124158-20b82306-0d61-41d8-8fa1-7ec2376608b5.png)

![clear-2](https://user-images.githubusercontent.com/1761249/196124175-9c18d405-0735-4a1a-8975-3671ddb24f91.png)

**Reply by dodyg** [reply]
Here's the weird part 

- `shape_clear_alternates` and `shape_add_alternates` work fine in clearing alternates and adding one alternate. 
- `shape_render` brings back all the previously cleared alternates plus the newly added one! 

![render-1](https://user-images.githubusercontent.com/1761249/196133502-bb30cb24-1a1a-4a54-94d7-1d92334123f9.png)
![render-2](https://user-images.githubusercontent.com/1761249/196133518-897e672d-0989-44a8-9b5d-129ff9e11b24.png)

**Reply by ns8482e** [reply]
Correct! Part's alternates are added during OnDisplaying event which is called before rendering a shape.  Here You are clearing/adding alternates of part's in content's view - ie. before the Part's `OnDisplaying` event. 

When you call `shape_render` for `Model.Content`, `OnDisplaying` is called for all shape items within `Model.Content` and that event adds alternates that appended in the alternate list. 

The priority order : Last in the list of alternates gets higher priority and rendered if that exists.

**Reply by dodyg** [reply]
Well the thing I want them to just use **one** alternate that I specify. How can I achieve this?

**Reply by ns8482e** [reply]
I mean if there are OOTB alternates are added in the list but they won't render as they don't exists. (until you define them in templates) 

It will render only last available template from alternate lists

**Reply by dodyg** [reply]
OK that's clear. Thanks. I have another follow up question regarding `BagPart`.

I have created this template and it got picked up and rendered.

![shape-1](https://user-images.githubusercontent.com/1761249/196216553-ba70e907-5c63-4f32-b39f-25d59fe1bafe.png)

The problem is I have no idea how to access the data to display it in the template because `{{Model.ContentItem}}` doesn't exist. 

![shape-2](https://user-images.githubusercontent.com/1761249/196216739-76f9fb04-e49b-41e9-aabf-bc700e56284d.png)

**Reply by ns8482e** [reply]
Availability of `ContentItem` property is depends on shape template being rendered and it's model

Here you are in template of typed part (`BagPart`) and the shape `HomePage_Detail__BagPart`. The Model that is represented by shape is `BagPartViewModel` [see](https://github.com/OrchardCMS/OrchardCore/blob/3dacb3412ffa55e397fff92e7831391f6e4b44a0/src/OrchardCore.Modules/OrchardCore.Flows/ViewModels/BagPartViewModel.cs#L13) ,  Here `ContentItem` can be accessed using `Model.BagPart.ContentItem`

**Reply by dodyg** [reply]
OK so 

`{{ Model.BagPart.Content }}` returns the content of the BagPart itself.
`{{ Model.BagPart.ContentItem }}` returns the whole `ContentItem` from the parent.

![content](https://user-images.githubusercontent.com/1761249/196419174-1fe88518-822b-4766-b584-2a9ff21291ea.png)

![content-2](https://user-images.githubusercontent.com/1761249/196419204-58e5908c-871a-4152-adbb-361c59e40017.png)

**Reply by dodyg** [reply]
`{{ Model }}` returns the shape of the BagPart

![shape-1](https://user-images.githubusercontent.com/1761249/196419697-e0aadef1-c97d-41b7-a480-1e36255a7b92.png)
![shape-2](https://user-images.githubusercontent.com/1761249/196419706-75cbb6e9-f030-444e-aa86-bd92224df0cd.png)

**Reply by dodyg** [reply]
@ns8482e  My question is how do I access the shape of the items contained in the `BagPart`?

**Reply by ns8482e** [reply]
That’s the original answer, Use [shape_build_display](https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#shape_build_display) for each item, 

The default implementation of BagPart creates shapes in view, so if you are overriding default then you  also need to build shapes for each item

**Comment by ns8482e** [comment]
@dodyg From the docs https://docs.orchardcore.net/en/latest/docs/reference/modules/Flow/BagPart/#templating-with-display-management

**Reply by dodyg** [reply]
I read that part. I don't understand it. Let me try.

**Reply by ns8482e** [reply]
The model is BagPartViewModel [see](https://github.com/OrchardCMS/OrchardCore/blob/3dacb3412ffa55e397fff92e7831391f6e4b44a0/src/OrchardCore.Modules/OrchardCore.Flows/ViewModels/BagPartViewModel.cs#L13) , which has `ContentItems` property

The default implementation of BagPart view https://github.com/OrchardCMS/OrchardCore/blob/3dacb3412ffa55e397fff92e7831391f6e4b44a0/src/OrchardCore.Modules/OrchardCore.Flows/Views/BagPart.cshtml#L7-L12

**Reply by dodyg** [reply]
@ns8482e  so now this works. However I don't understand why `Detail` in ` {{ item | shape_build_display: "Detail" | shape_render }}` is so special. If I modify `Detail` to something else, it will generate error.

![working-1](https://user-images.githubusercontent.com/1761249/196426247-42d7f169-9bb8-43e7-b246-44ecf40d588e.png)

![working-2](https://user-images.githubusercontent.com/1761249/196426265-624f3493-c9ba-4d78-b337-d0428802aca3.png)

**Reply by dodyg** [reply]
Ok perhaps `DisplayType:Detail` is a very specific value. The problem here right now is that the Alternate is `Content__Paragraph`. The next logical step is to add another alternate via `shape_add_alternates` so we can provide a custom template of the content.

![the-new-shape](https://user-images.githubusercontent.com/1761249/196427537-d7264888-d781-4899-9a11-51fe770cfdc5.png)

![the-new-shape-2](https://user-images.githubusercontent.com/1761249/196427548-f41c682a-dc10-414e-a918-b7a378105727.png)

**Reply by dodyg** [reply]
OK for `BagPart` `Model.ContentItems` also works

![content-items-1](https://user-images.githubusercontent.com/1761249/196428320-aa63fc5f-7f7a-4746-aeff-eef9c414f640.png)

![content-items-2](https://user-images.githubusercontent.com/1761249/196428324-0c3a075d-b67e-45ca-86bf-b3ee305a1776.png)

**Reply by dodyg** [reply]
We can assign a specific alternate for the item contained within `BagPart` inside `HomePage` Content Type this way.

![custom-1](https://user-images.githubusercontent.com/1761249/196430739-89e30c05-183a-4f3e-84ac-fc41734242c1.png)

![custom-2](https://user-images.githubusercontent.com/1761249/196430749-573ffb3d-6cbb-4045-8a88-0b72fdbbbc69.png)

**Reply by ns8482e** [reply]
Yes Detail is DisplayType. In  ootb it’s based on your settings of BagPart

**Reply by dodyg** [reply]
@ns8482e thank your patient and for all your help. 

For others, this is the complete example on how to render a `BagPart`.

This is the final result. We specify a custom alternate to render instead of using the ones suggested by OC.

![result-1](https://user-images.githubusercontent.com/1761249/196436656-1ca20880-d87f-483e-b80f-7a1cbc9f0453.png)

This is how to build the alternate name. You can use your own convention on how name your alternate for this purpose.

![result-2](https://user-images.githubusercontent.com/1761249/196436668-528edbe8-35f2-4140-8ed6-e9995db072ff.png)

This is the details of this alternate. 

![result-3](https://user-images.githubusercontent.com/1761249/196436670-753d0d81-d29e-4bb3-ae7a-55c43b7b7b64.png)
